### ISSUE TITLE
Improvement in releasing local shard files post relocation.

### ISSUE THREAD

**Issue by dhwanilpatel** [opensearch-project/OpenSearch#711]
### **Is your feature request related to a problem? Please describe.**
If index has multiple replicas configured and shards are getting relocated to another node then if relocation of some primary/replica shard is completed and other replica is still being relocated then local shard files of this shards on source node of completed replica is not getting deleted until other relocation are also not completed.

For e.g. 
Index A has 1 shard and 2 replica configured. (p, r1, r2). Now all shards are getting relocated p ( from n1 -> m1), r1 ( from n2 -> m2), r3 ( from n3 -> m3). 
Relocation of p and r1 is completed and r2 is still in progress. Since relocation of r2 is in progress local shard files on n1/n2 nodes are not getting deleted until this relocation is not completed. 

This is because of strategy of removing local shard files, it verifies all the shard copy should be active on other nodes before deleting local shard files. ([code pointer](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/IndicesStore.java#L220-L237)). 

### **Describe the solution you'd like**
There is one validation before performing above strategy where it relies on the Shard State from ShardRoutingTable. ([Code pointer](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/IndicesStore.java#L207)) 
There it is just checking for started shard only not active shards(i.e. started + relocating). We can more flexible in validation and include relocating shards as well. As for relocating shards shard would be in active on source node. For above strategy source node of relocating shard will sent ack as shard is active on source node.

### **Describe alternatives you've considered**
I have considered to change the strategy of waiting until all shard copy to be in STARTED before deleting shard files to be more optimistic and rely on only one copy to be active. 
It will lead to more costly recovery post node drops as after node drops there will be peer recovery for replicas instead of local store recovery.

**Comment by itiyamas** [comment]
Hi Dhwanil,

Can you provide steps to reproduce this issue?
Would be good to document the issues caused due to not deleting the local copy of the shard.

**Comment by dhwanilpatel** [comment]
Hello @itiyamas 

Thanks for your suggestion. I have updated issue with it.

**Comment by itiyamas** [comment]
This is a nice improvement, but we need to be more careful about this change as recovery as it involves wiping the data for shards and can potentially result in data loss in extreme edge cases. Some points to consider here:

1. A replica always happens from primary and replica may be recovering from the old primary when we delete this data. I am assuming that recovery will be cancelled when you delete the data from disk, but we need to understand the failure modes better here.
2. Additionally, the state movements of primary and replica on data nodes are via the asynchronous cluster state applier, which is when the individual recoveries are cancelled- the speed of applying this state may be different at different nodes. Hence, we may need to keep the data around for some time for preventing read/write failures.
3. Similarly for replica, it may have been a primary from which a recovery was happening. Once the primary switches state to replica - the old recovery from replica will be cancelled and when the replica finally moves to a new node. 

Can we add more tests for all edge cases and observe the system behavior. We could delay the application of cluster state across nodes and check the impact?
Also, let us delete the data for old replicas first. Once it is stable, we can make this change for primaries.

**Comment by dhwanilpatel** [comment]
> 1. A replica always happens from primary and replica may be recovering from the old primary when we delete this data. I am assuming that recovery will be cancelled when you delete the data from disk, but we need to understand the failure modes better here.

Yes, it will trigger new recovery for replica but it is not because of deletion of data. In existing behavior as well (where data of primary is not deleted post successful relocation),  it triggers the new recovery for replica if the state of primary is changed.

>2. Additionally, the state movements of primary and replica on data nodes are via the asynchronous cluster state applier, which is when the individual recoveries are cancelled- the speed of applying this state may be different at different nodes. Hence, we may need to keep the data around for some time for preventing read/write failures.

This shards files are also deleted at the time of cluster state applier only. Before deleting files, ES sends request to each node and get acknowledgment that shard is started on that node, post that it will delete the data. This will take care of scenario  where different nodes are yet to apply cluster state. File deletion logic will wait until it gets the acknowledgment.

>3. Similarly for replica, it may have been a primary from which a recovery was happening. Once the primary switches state to replica - the old recovery from replica will be cancelled and when the replica finally moves to a new node.

Yes, when replica is recovery is happening from primary and state of primary changes like relocates to different node then recovery of replica would be cancelled and new recovery would start from new node.

**We are not suggesting any new mechanism for deleting shard files**. we are keeping the mechanism as it is. 

Mechanism is very conservative, it deletes the shard files from local disk if **all the copies are STARTED** on other nodes. 
For this mechanism ES sends the requests different nodes and get the ack back and then removes the data. ([code pointer](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/IndicesStore.java#L220-L237)). 

ES had pre checks before executing above mechanism([code pointer](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/IndicesStore.java#L164)), so it don't have to send this requests to nodes every time there is new cluster state. This validation tries to check the state of the shard from shard routing table. If this validation passes then they perform above mechanism and if it satisfies then it deletes the data. 

There seems to be bug, in this precheck as it checks if all the shards are STARTED or not.([code pointer for pre check](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/IndicesStore.java#L207)) Here we can add RELOCATING shard as well, as for RELOCATING shard, shard is active and performing read/writes on the source node from where shard is getting relocated. I have added RELOCATING state as well in pre check. 

I am changing just pre-check method, not mechanism for deleting data. I have raised PR for this(#742 ), can you please take a look at that? May be from there it would be easier to understand the change I am proposing. 

> Can we add more tests for all edge cases and observe the system behavior. We could delay the application of cluster state across nodes and check the impact?

ES already have integration test for this behavior by adding delay/block in application of cluster state.([code pointer](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/indices/store/IndicesStoreIntegrationIT.java)). Those integ tests are passing with my change in precheck method.

**Comment by itiyamas** [comment]
> Yes, it will trigger new recovery for replica but it is not because of deletion of data. In existing behavior as well (where data of primary is not deleted post successful relocation), it triggers the new recovery for replica if the state of primary is changed.

I am not talking about when recovery will happen for a replica, but how will existing replica recoveries be impacted. 

>This shards files are also deleted at the time of cluster state applier only. Before deleting files, ES sends request to each node and get acknowledgment that shard is started on that node, post that it will delete the data. This will take care of scenario where different nodes are yet to apply cluster state. File deletion logic will wait until it gets the acknowledgment.

The nodes can have different views and there could be edge cases where different nodes have different views about cluster state as application may happen at different speeds. In some cases, a node may be completely lagged.

>Yes, when replica is recovery is happening from primary and state of primary changes like relocates to different node then recovery of replica would be cancelled and new recovery would start from new node.

Can we test this case with delayed cluster state application.

A few other points to be noted:
1. 1 shard with only 1 primary and no replica. An inactive primary relocates to a new node. Now, new node goes down before old data could be deleted. Does the old primary get promoted to the primary. I think it should be as sync id markers should be the same. Now, we are changing that case as well with the new change and we need to think through failure modes around it and add tests accordingly, if they do not exist.
2. Any bug in the allocator framework that messes up with the nodes info can result in data to be deleted more aggressively now.

My two cents here would be to make this change for replicas only. However, let us confirm that tests exist for all the edge cases that we talked about here before making this change for primaries.

**Comment by dhwanilpatel** [comment]
Thanks @itiyamas  for clarification, you are correct there is edge cases between recovery/deletion which we need to fix before making this change for primaries. 

When primary is relocated to different node, cancellation of other replica recovery from old primary is being performed via [`IndicesClusterStateService`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java#L229) and deletion of files are being performed via [`IndicesStore`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/IndicesStore.java#L136). These both operation are being performed via different cluster state listener, so they would be executed in any order. There is chance that we delete the shard files of old primaries before the recoveries are cancelled, that can cause recovery to be failed. 

I see there is [TODO](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/IndicesStore.java#L90-L91) in `IndicesStore` listener to move it `IndicesClusterStateService`, post that we can define some order of above operations in `IndicesClusterStateService`. Maybe we can do that, but for that we will need to evaluate behavior of other operation(if any) of `IndiceStore` before doing refactoring. 

**As per your suggestion, Right now I will make changes for replicas only.** 

Couple of points to highlight in regards of some of edgecase whenever we want to think for making this change for primary,

> A replica always happens from primary and replica may be recovering from the old primary when we delete this data. I am assuming that recovery will be cancelled when you delete the data from disk, but we need to understand the failure modes better here.

As discussed above, we need to fix this before making this change for primary.

> The state movements of primary and replica on data nodes are via the asynchronous cluster state applier, which is when the individual recoveries are cancelled- the speed of applying this state may be different at different nodes. Hence, we may need to keep the data around for some time for preventing read/write failures.

Recoveries are being cancelled from the source node, and target node performs the retries of the recovery. Once primary is relocated to other node, and when cluster state applier runs on old primary node it will cancel all recoveries. In this case different nodes will perform retries at different time as per node's speed. I think this should be fine.

> 1 shard with only 1 primary and no replica. An inactive primary relocates to a new node. Now, new node goes down before old data could be deleted. Does the old primary get promoted to the primary. I think it should be as sync id markers should be the same. Now, we are changing that case as well with the new change and we need to think through failure modes around it and add tests accordingly, if they do not exist.

When primary is relocated to different node, `IndicesClusterStateService` close that shard on old node. As per my understanding closed shard wouldn't be promoted for active primary even if shards files are not deleted. Please correct me if I am wrong here.

**Comment by itiyamas** [comment]
I agree with making this change incrementally for replicas first. Just be cautious of cases when replica becomes primary during ongoing recoveries and ensure that tests exist- both for cases when indexing is ongoing and when there is no indexing- basically same and different sync id markers.
